README.md
# routing-py

[![Latest
Version](https://img.shields.io/pypi/v/routing-py.svg?style=flat-square)](https://pypi.python.org/pypi/routing-py/)

[![Build
Status](https://img.shields.io/travis/routing-py/routing-py.svg?style=flat-square)](https://travis-ci.org/gis-ops/routing-py)

[![License](https://img.shields.io/github/license/routing-py/routing-py.svg?style=flat-square)](https://pypi.python.org/pypi/routing-py/)

*One to unite them all* - routing-py is a Python 3 client for several popular routing webservices.

This library makes it extremely easy for Python developers to request directions, isochrones or time-distance matrices using third-party spatial webservices.

routing-py currently includes support for the following services:

- [Mapbox, either Valhalla or OSRM](https://docs.mapbox.com/api/navigation)
- [Openrouteservice](https://openrouteservice.org/dev/#/api-docs)
- [Here Maps](https://developer.here.com/documentation)
- [Google Maps](https://developers.google.com/maps/documentation)
- [Graphhopper](https://graphhopper.com/api/1/docs)
- [Local Valhalla](https://github.com/valhalla/valhalla-docs)
- [Local Mapbox](https://github.com/Project-OSRM/osrm-backend/wiki)

The up-to-date list is available on the [routing-py doc section](https://routing-py.readthedocs.io/en/latest/#routers). 
Router classes are located in [routing-py.routers](https://github.com/gis-ops/routing-py/tree/master/routing-py/routers).

routing-py is tested against versions 3.6, 3.7. 

© routing-py contributors 2018 under the [Apache 2.0 License](https://github.com/gis-ops/routing-py/blob/master/LICENSE).

## Installation

Install using [pip](http://www.pip-installer.org/en/latest/) with:

    pip install routing-py

Or, [download a wheel or source archive from PyPI](https://pypi.python.org/pypi/routing-py).

## Generic Response Objects

...

## Directions using openrouteservice vs. Google Maps

To geolocate a query to an address and coordinates:

``` python
>>> from geopy.geocoders import Nominatim
>>> geolocator = Nominatim(user_agent="specify_your_app_name_here")
>>> location = geolocator.geocode("175 5th Avenue NYC")
>>> print(location.address)
Flatiron Building, 175, 5th Avenue, Flatiron, New York, NYC, New York, ...
>>> print((location.latitude, location.longitude))
(40.7410861, -73.9896297241625)
>>> print(location.raw)
{'place_id': '9167009604', 'type': 'attraction', ...}
```

## Isochrones using Mapbox-Valhalla vs. HERE Maps

To geolocate a query to an address and coordinates:

``` python
>>> from geopy.geocoders import Nominatim
>>> geolocator = Nominatim(user_agent="specify_your_app_name_here")
>>> location = geolocator.reverse("52.509669, 13.376294")
>>> print(location.address)
Potsdamer Platz, Mitte, Berlin, 10117, Deutschland, European Union
>>> print((location.latitude, location.longitude))
(52.5094982, 13.3765983)
>>> print(location.raw)
{'place_id': '654513', 'osm_type': 'node', ...}
```

## Matrix using Mapbox-OSRM vs Google Maps

To find the address corresponding to a set of coordinates:

``` python
>>> from geopy.geocoders import Nominatim
>>> geolocator = Nominatim(user_agent="specify_your_app_name_here")
>>> location = geolocator.reverse("52.509669, 13.376294")
>>> print(location.address)
Potsdamer Platz, Mitte, Berlin, 10117, Deutschland, European Union
>>> print((location.latitude, location.longitude))
(52.5094982, 13.3765983)
>>> print(location.raw)
{'place_id': '654513', 'osm_type': 'node', ...}
```

## Contributions

...


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

**Unreleased** is available in Github's `master` branch, but not on PyPI.

## **Unreleased**

### Fixed

- Fixes taking into account the `preference` parameter when calculating isochrones and matrix with Valhalla ([#120](https://github.com/gis-ops/routingpy/issues/120))
- Google's matrix checks each response element's status code [#122](https://github.com/gis-ops/routingpy/pull/122)

## [v1.3.0](https://pypi.org/project/routingpy/1.3.0/)

### Added
- Valhalla `/expansion` examples in the Jupyter Notebook
- OpenTripPlanner v2 support for routing & isochrones

### Fixed
- Google router's `duration` and `distance` attributes not being calculated correctly in single route response ([#107](https://github.com/gis-ops/routingpy/issues/107))
- pointer issue when requesting multiple times with the same connection

### Removed
- `MapboxValhalla` provider, since Mapbox doesn't expose a public Valhalla endpoint anymore

## [v1.2.1](https://pypi.org/project/routingpy/1.2.1/)
### Fixed
- Graphhopper 7.0 deprecated the "vehicle" parameter for the new "profile" parammeter


## [v1.2.0](https://pypi.org/project/routingpy/1.2.0/)
### Fixed
- Unit conversion did not work properly in ors' directions method
- Docstring for `intersections` parameter in ors' isochrones method was missing
- `profile` parameter was unnecessarily passed to POST params in ors' isochrones and matrix methods
- switched Graphhopper to POST endpoints & fixed any outdated parameters

## [v1.1.0](https://pypi.org/project/routingpy/1.1.0/)

### Added
- Added Valhalla's trace_attributes endpoint

## [v1.0.4](https://pypi.org/project/routingpy/1.0.4/)

### Fixed
- Passing a Valhalla.Waypoint to isochrones resulted in an unhandled exception

## [v1.0.3](https://pypi.org/project/routingpy/1.0.3/)

### Fixed
- OSRM does have a weird URL concept for profiles, so revert [#64](https://github.com/gis-ops/routing-py/issues/64))

### Added
- `interval_type` to `Isochrone` and `Expansions` objects

## [v1.0.2](https://pypi.org/project/routingpy/1.0.2/)

### Fixed
- Valhalla's 'matrix' endpoint couldn't deal with NULL entries

## [v1.0.0](https://pypi.org/project/routingpy/1.0.0/)

### Changed
- made all `parse_*` functions public so they can be used by super projects

### Fixed
- OSRM wasn't requesting the right endpoints: profile is meaningless
- GraphHopper parsing fails with not encoded points ([#54](https://github.com/gis-ops/routing-py/issues/54))
- Allow "narrative" argument for Valhalla's directions endpoint

## [0.4.0](https://pypi.org/project/routingpy/0.4.0/)

### Added
- support for Valhalla's `/expansion` endpoint
- support for HereMaps kwargs arguments

### Fixed
- enhanced error handling for directions with Google [#15](https://github.com/gis-ops/routing-py/issues/15)
- fixed MapboxOSRM client, removing unused parameters and unifying Isochrones interface [#21](https://github.com/gis-ops/routing-py/issues/21)
- fixed bug that caused HereMaps client to use wrong API endpoints [#28](https://github.com/gis-ops/routing-py/pull/28)
- fixed bug that caused HereMaps Isochrones to return lists of `list_reverseiterator`s instead of coordinates [#29](https://github.com/gis-ops/routing-py/issues/29)
- updated jupyter notebooks examples
- switched coordinate order of OSRM
- if no get_params are defined, omit "?" from the request URL

### Changed

- BREAKING: pulled client stuff into a separate Client class which gets passed to the individual router instances with the default being the same as before [#24](https://github.com/gis-ops/routing-py/pull/24)


## [0.3.3](https://github.com/gis-ops/routing-py/releases/tag/0.3.3)
### Added
- 2020/2021 Valhalla HTTP API improvements
### Fixed
- README local OSRM description

## [0.3.2](https://github.com/gis-ops/routing-py/releases/tag/0.3.2)
### Fixed
- HERE isochrones had lat, lon instead of lon, lat ([#14](https://github.com/gis-ops/routing-py/issues/14))

## [0.3.1](https://github.com/gis-ops/routing-py/releases/tag/0.3.2)
### Fixed
- HERE routers can now also be used with api key
- HERE isochrones had lat, lon instead of lon, lat ([#14](https://github.com/gis-ops/routing-py/issues/14))
- Set Google router's profile queryparam correctly (was set to "profile" now is "mode")

## [0.3.0](https://github.com/gis-ops/routing-py/releases/tag/0.3.0) 2020-08-09
### Added
- GeoJSON support for Graphhopper's isochrones
- [MyBinder](https://mybinder.org/v2/gh/gis-ops/routing-py/master?filepath=examples) notebook collection
### Fixed
- Add departure and arrival parameter for HERE isochrones API
- OSRM Mapbox isochrone ranges were using floats (#2)
- OSRM matrix ouputs distances array now (#6)
- Graphhopper isochrones used wrong vehicle parameter
### Changed
- Profile parameter for HERE behaves now like other routers
- ORS alternative_routes parameter (#4)
- Brought Graphhopper support to its v1.x release
- Minimum Python version > 3.6.1 (for Pandas 1.x)
### Deprecated
-

## [0.2](https://github.com/gis-ops/routing-py/releases/tag/v0.2) 2019-04-30
### Added
- Example notebook to compare providers
### Fixed
- Here matrix did not allow `sources` and `destinations` to be optional
- Base class used sorted(params) which messed up the order of parameters for Graphhopper endpoints
- MapboxOSRM directions was parsing geometry to \[lat, lon\]
- README not valid on PyPI, wasn't rendered
- add to PyPI and Conda
### Changed
-
### Deprecated
-

## [0.1 - First Release](https://github.com/gis-ops/routing-py/releases/tag/v0.1) 2019-04-14


misc.py
from routingpy import ORS, Graphhopper, HereMaps

#api = ORS(key='5b3ce3597851110001cf624858ff1f9270364f74907d578b80c3c41cc', base_url='https://api.openrouteservice.org', requests_kwargs={})

#routes = api.directions(coordinates=[[8.681495, 49.41461], [8.686507, 49.41943]], profile='driving-car', format='json', dry_run=False)

#print(routes)

# from routingpy import Google

# api = Google(api_key='AIzaSyBibIkPFip29wcE4y4wrZnY_mnwZv_sU8g')

# routes = api.directions(
#     coordinates=[[8.688641, 49.420577], [8.680916, 49.415776],
#                  [8.780916, 49.445776]],
#     profile='car',
#     dry_run=False)

# print(routes.distance)

# from routingpy import Graphhopper

# api = Graphhopper(api_key='78a38b3d-bd4f-4e52-b466-37053a0344fa')

# routes = api.directions(
#     coordinates=[[8.688641, 49.420577], [8.680916, 49.415776],
#                  [8.780916, 49.445776]],
#     elevation=False,
#     # points_encoded=True,
#     profile='car',
#     # avoid=['motorway', 'toll'],
#     format='json',
#     # instructions=False,
#     dry_run=False)

# print(routes.geometry, routes.duration, routes.distance)

# isochrone = api.isochrones(
#     coordinates=[8.34234, 48.23424],
#     range=[200],
#     profile='car',
#     buckets=3,
#     dry_run=True)

# print(isochrone)

# #matrix = api.distance_matrix(coordinates=[[11.588051, 49.932707], [10.747375,50.241935], [11.983337,50.118817]], out_array=['weights', 'times', 'distance'], profile='car', dry_run=False)

# #print(matrix)

# # matrix1 = api.distance_matrix(
# #     coordinates=[[8.688641, 49.420577], [8.680916, 49.415776],
# #                  [8.780916, 49.445776]],
# #     sources=[0, 1],
# #     destinations=[2],
# #     out_array=['weights', 'times', 'distance'],
# #     profile='car',
# #     dry_run=False)
# #matrix1 = api.distance_matrix(coordinates=[[8.688641, 49.420577], [8.680916, 49.415776], [8.780916, 49.445776]], sources=[0,1], out_array=['weights', 'times', 'distance'], profile='car', dry_run=False)

# #print(matrix1)

# from routingpy import HereMaps
# import datetime

here_api = HereMaps(
    app_id='Ssw36q9nNyfx6899yBPK', app_code='L6SDa5v0jzsgRr1yAtMWFw')

# # routes = here_api.directions(
# #     # coordinates=[
# #     #     [8.688641, 49.420577],
# #     #     [8.680916, 49.415776],
# #     #     [8.780916, 49.445776]
# #     # ],
# #     coordinates=[
# #         here_api.WayPoint(
# #             position=(8.688641, 49.420577),
# #             waypoint_type='stopOver',
# #             stopover_duration=30,
# #             transit_radius=200,
# #             heading=180),
# #         here_api.WayPoint(
# #             position=(8.680916, 49.415776),
# #             waypoint_type='passThrough',
# #             user_label='honk',
# #         ),
# #         here_api.WayPoint(position=(8.780916, 49.445776))
# #     ],
# #     profile='car;fastest',
# #     # profile=here_api.RoutingMode(
# #     #     mode_type='fastest',
# #     #     mode_transport_type='car',
# #     #     mode_traffic='enabled',
# #     #     features={
# #     #         "motorway": -2,
# #     #         "tollroad": -1,
# #     #         #"tunnel": 1
# #     #     }),
# #     request_id=101,
# #     avoid_areas=[[(8.688641, 49.420577), (8.680916, 49.415776)],
# #                  [(8.780916, 49.445776), (8.780916, 49.445776)]],
# #     #avoid_links=[-53623477, -536234327, 123],
# #     avoid_seasonal_closures=True,
# #     #avoid_turns='difficult',
# #     allowed_zones=[167772378],
# #     exclude_zones=[510, 511],
# #     exclude_zone_types=['vignette', 'congestionPricing'],
# #     exclude_countries=['AUT', 'CHE'],
# #     departure=datetime.datetime(2019, 3, 29, 3, 0).isoformat(),
# #     #arrival=datetime.datetime(2019, 3, 29, 3, 0).isoformat(),
# #     alternatives=3,
# #     metric_system='metric',
# #     view_bounds=[(8.688641, 49.420577), (8.680916, 49.415776)],
# #     resolution={
# #         'viewresolution': 300,
# #         'snapresolution': 300
# #     },
# #     instruction_format='text',
# #     language='en-us',
# #     route_attributes=[
# #         'waypoints', 'summary', 'summaryByCountry', 'shape', 'boundingBox',
# #         'legs', 'notes', 'lines', 'routeId', 'groups', 'tickets', 'incidents',
# #         'zones'
# #     ],
# #     leg_attributes=['maneuvers', 'waypoint', 'length', 'travelTime'],
# #     maneuver_attributes=['position', 'length', 'travelTime'],
# #     link_attributes=['shape', 'speedLimit'],
# #     dry_run=False)

# #print(routes.__repr__)

# isochrones = here_api.isochrones(
#     # coordinates=
#     #       here_api.WayPoint(
#     #           position=(8.688641, 49.420577)
#     #       )
#     # ,
#     coordinates=[8.34234, 48.23424],
#     ranges=[3000],
#     range_type='distance',
#     profile='car;fastest',
#     #center_type='start',
#     #request_id=101,
#     #single_component=False,
#     #quality=1,
#     dry_run=False
#     #max_points=100,

#     # profile=here_api.RoutingMode(
#     #     mode_type='fastest',
#     #     mode_transport_type='car',
#     #     mode_traffic='enabled',
#     #     features={
#     #         "motorway": -2,
#     #         "tollroad": -1,
#     #         #"tunnel": 1
#     #     }),
#)

#print(isochrones[0].geometry, isochrones[0].center, isochrones[0].range )

ors_api = ORS(
    api_key='5b3ce3597851110001cf624858ff1f9270364f74907d578b80c3c41cc')

isochrones = ors_api.isochrones(
    coordinates=[[8.34234, 48.23424]],
    intervals=[3000],
    interval_type='distance',
    profile='driving-car',
    dry_run=False)

print(isochrones, isochrones[0].geometry, isochrones[0].center,
      isochrones[0].range)

# #print(isochrones.__repr__)
# #print(isochrones.isochrones[2].__repr__)

# matrix = here_api.distance_matrix(
#     coordinates=[[8.688641, 49.420577], [8.680916, 49.415776],
#                  [8.780916, 49.445776]],
#     sources=[0, 1, 2],
#     destinations=[0, 1, 2],
#     profile='car;fastest',
#     summary_attributes=['traveltime', 'costfactor', 'distance'],
#     truck_type='truck',
#     trailers_count=3,
#     shipped_hazardous_goods=['gas', 'flammable'],
#     limited_weight=10,
#     weight_per_axle=100,
#     height=20,
#     width=10,
#     length=10,
#     # coordinates=[
#     #     here_api.WayPoint(
#     #         position=(8.688641, 49.420577),
#     #         waypoint_type='stopOver',
#     #         stopover_duration=30,
#     #         transit_radius=200,
#     #         heading=180),
#     #     here_api.WayPoint(
#     #         position=(8.680916, 49.415776),
#     #         waypoint_type='passThrough',
#     #         user_label='honk',
#     #     ),
#     #     here_api.WayPoint(position=(8.780916, 49.445776))
#     # ],
#     #search_range=10000,
#     dry_run=True

#     # profile=here_api.RoutingMode(
#     #     mode_type='fastest',
#     #     mode_transport_type='car',
#     #     mode_traffic='enabled',
#     #     features={
#     #         "motorway": -2,
#     #         "tollroad": -1,
#     #         #"tunnel": 1
#     #     }),
# )

# # #print(matrix.__repr__)

# routes = here_api.directions(
#     # coordinates=[
#     #     here_api.WayPoint(
#     #         position=(8.688641, 49.420577),
#     #         waypoint_type='stopOver',
#     #         stopover_duration=30,
#     #         transit_radius=200,
#     #         heading=180),
#     #     here_api.WayPoint(
#     #         position=(8.680916, 49.415776),
#     #         waypoint_type='passThrough',
#     #         user_label='honk',
#     #     ),
#     #     here_api.WayPoint(position=(8.780916, 49.445776))
#     # ],
#     # routing_mode=here_api.RoutingMode(
#     #     mode_type='fastest',
#     #     mode_transport_type='car',
#     #     mode_traffic='enabled',
#     #     features={
#     #         "motorway": -2,
#     #         "tollroad": -1,
#     #         "tunnel": 1
#     #     }),
#     # allowed_zones=[167772378],
#     # custom_consumption_details='',
#     # line_attributes=['lineForeground', 'lineBackground'],
#     # walk_time_multiplier=2.0,
#     # walk_speed=1.2,
#     # walk_radius=2000,
#     # representation=[
#     #    'overview', 'display', 'dragNDrop', 'navigation', 'linkPaging',
#     #    'turnByTurn'
#     # ],
#     # tunnel_category=['A'],
#     coordinates=[(8.688641, 49.420577), (8.680916, 49.415776),
#                  (8.780916, 49.445776)],
#     profile='truck;fastest',
#     request_id=101,
#     avoid_areas=[[(8.688641, 49.420577), (8.680916, 49.415776)],
#                  [(8.780916, 49.445776), (8.780916, 49.445776)]],
#     avoid_links=[-53623477],
#     avoid_seasonal_closures=True,
#     avoid_turns='difficult',
#     exclude_zones=[510, 511],
#     exclude_zone_types=['vignette', 'congestionPricing'],
#     exclude_countries=['AUT', 'CHE'],
#     departure=datetime.datetime(2019, 3, 29, 3, 0).isoformat(),
#     alternatives=3,
#     metric_system='metric',
#     view_bounds=[(8.688641, 49.420577), (8.680916, 49.415776)],
#     resolution={
#         'viewresolution': 300,
#         'snapresolution': 300
#     },
#     instruction_format='text',
#     language='en-us',
#     json_attributes=9,
#     route_attributes=[
#         'waypoints', 'summary', 'summaryByCountry', 'shape', 'boundingBox',
#         'legs', 'notes', 'lines', 'routeId', 'groups', 'tickets', 'incidents',
#         'zones'
#     ],
#     leg_attributes=['maneuvers', 'waypoint', 'length', 'travelTime'],
#     maneuver_attributes=['position', 'length', 'travelTime'],
#     link_attributes=['shape', 'speedLimit'],
#     generalization_tolerances=[0.1, 0.01],
#     vehicle_type='diesel,5.5',
#     license_plate='lastcharacter:5',
#     max_number_of_changes=5,
#     combine_change=False,
#     truck_type='truck',
#     trailers_count=3,
#     shipped_hazardous_goods=['gas', 'flammable'],
#     limited_weight=10,
#     weight_per_axle=100,
#     height=20,
#     width=10,
#     length=10,
#     truck_restriction_penalty='soft',
#     return_elevation=True,
#     consumption_model='default',
#     speed_profile='fast',
#     dry_run=False)

# print(routes)
